### DISCUSSION TITLE
Sharing data among multiple instances of same Job type.

### DISCUSSION THREAD

**Discussion by Jalalx** [quartznet/quartznet#487] [General]
Hello guys.
Is it OK to share some data among multiple instances of a Job using `ConcurrentQueue<T>` like this?
```
    public class MediaTrackerJob : IJob
    {
        private readonly static object SyncLock = new object();
        private readonly ConcurrentQueue<Media> CachedMedia = new ConcurrentQueue<Media>();

        private readonly IStorageService _storageService = new DapperStorageService();

        public MediaTrackerJob()
        {
        }

        public void Execute(IJobExecutionContext context)
        {
            MakeSureCacheIsNotEmpty();

            Media media;
            if (CachedMedia.TryDequeue(out media))
            {
                // some code goes here...
                media.LastUpdateDate = DateTime.Now;
                _storageService.SaveMedia(media);
            }
        }

        private void MakeSureCacheIsNotEmpty()
        {
            if (CachedMedia.IsEmpty)
            {
                lock (SyncLock)
                {
                    if (CachedMedia.IsEmpty)
                    {
                        // ORDER BY media.LastUpdateDate ASC;
                        var mediaList = _storageService.GetRecentMediaForUpdate(AppSettings.AgentId);
                        foreach (var media in mediaList)
                        {
                            CachedMedia.Enqueue(media);
                        }
                    }
                }
            }
        }
    }
```
What are consequences of this pattern? Is there any better solution?

**Comment by lahma** [comment]
You should generally avoid this, it's not completely wrong to do so and it will most likely work, but it's always have good separation of concerns and responsibilities. It almost sounds like you are trying to work around some other issue, like concurrent firing of same job (DisalllowConcurrentExcecutionAttribute would handle this).

I personally would just make sure there's single job for this purpose and disallowed to run concurrently, then periodic runs would just deplete work to do from DB.

**Comment by Jalalx** [comment]
@lahma So if I move caching logic to a separate class, it will be OK?

**Comment by lahma** [comment]
I'd rather like to know what you are trying to achieve. It seems that this is more like producer-consumer kind of a problem. You load data as fast as you can and try to process it.

* You could have two kind of jobs, one to load data and trigger jobs based on retrieved rows, triggering job with trigger job data map containing target information
* You could push to message queue via main job and use consumers outside of quartz to process as fast as possible (allows you to define how many consumers will be draining the queue)

I guess the main problem here is that you are trying to partition the work, not as easy when you don't have clear visibility on the worker counts, unless a dedicated scheduler with fixed thread pool size of 1 (job reading data and trigger jobs) + 
actual desired max worker count).

**Comment by eaba** [comment]
This is giving me some insights.
I currently have about 15 scheduler instances running different jobs. If am to run a main job with message queue and replace other instances with consumers and in other not to have bottleneck at the main queing job, how many transaction(triggers) per second can quartz.net handle in a single instance. Am thinking in the line of 'software pipeline'.

**Comment by lahma** [comment]
You really need to measure this yourself. It's such a "it depends", RAMJobStore vs AdoJobStore, job execution time...

**Comment by eaba** [comment]
I use AdoJobStore. By your records, what has been ur highest throughput? The configs involved too. 
Please!

**Comment by lahma** [comment]
Please measure by yourself. It really depends on job duration/resource usage, physical hardware etc. Generally Quartz can handle the thread pool size of concurrent jobs but the actual job execution time will determine the actual throughput.

**Comment by eaba** [comment]
Okay thanks
